ocaml/testsuite/tests/typing-modular-explicit/general.ml

```diff
@@ -0,0 +1,797 @@
+val typing_order2 :
+  ((module T/1 : Typ) -> T/1.t -> T/1.t) ->
+  ((module T/2 : Typ) -> T/2.t -> T/2.t) * Int.t = <fun>
+|}, Principal{|
+Line 1, characters 32-44:
+1 | let typing_order2 f = (id' f, f (module Int) 3)
+                                    ^^^^^^^^^^^^
+Warning 18 [not-principal]: this module packing is not principal.
+
+val typing_order2 :
+  ((module T/1 : Typ) -> T/1.t -> T/1.t) ->
+  ((module T/2 : Typ) -> T/2.t -> T/2.t) * Int.t = <fun>
+|}]
+
+(** The following test check that tests at module unpacking still happen with
+    modular explicits *)
+
+(* we test free type variables cannot occur *)
+module type T = sig type t val v : t end;;
+let foo (module X : T with type t = 'a) = X.v X.v;;
+
+[%%expect{|
+module type T = sig type t val v : t end
+Line 6, characters 16-17:
+6 | let foo (module X : T with type t = 'a) = X.v X.v;;
+                    ^
+Error: The type of this packed module contains variables:
+       (module T with type t = 'a)
+|}]
+
+(* Test principality warning of type *)
+let principality_warning2 f =
+  let _ : ((module T : Typ) -> T.t -> T.t) -> unit = f in
+  f (fun (module T) x -> x)
+
+[%%expect{|
+val principality_warning2 :
+  (((module T : Typ) -> T.t -> T.t) -> unit) -> unit = <fun>
+|}, Principal{|
+Line 3, characters 4-19:
+3 |   f (fun (module T) x -> x)
+        ^^^^^^^^^^^^^^^
+Warning 18 [not-principal]: this module unpacking is not principal.
+
+val principality_warning2 :
+  (((module T : Typ) -> T.t -> T.t) -> unit) -> unit = <fun>
+|}]
+
+(* This test check that as `t` is private we cannot inline its definition *)
+module type S = sig
+  type t = private int
+  val f : t
+end
+
+let check_escape : _ -> _ = fun (module M : S) -> M.f
+
+[%%expect{|
+module type S = sig type t = private int val f : t end
+Line 6, characters 50-53:
+6 | let check_escape : _ -> _ = fun (module M : S) -> M.f
+                                                      ^^^
+Error: This expression has type M.t but an expression was expected of type 'a
+       The type constructor M.t would escape its scope
+|}]
+
+(* The following test should give a warning only once.
+  Here we test that the structure is typed only once.
+  To achieve this we wrote a code that raises a warning. If the warning is raised
+  twice then that means typing happened twice.
+*)
+
+module type TInt = sig type t = int end
+
+let f (module T : TInt) (x : T.t) = x
+
+let raise_duplicated_warning =
+  f (module struct
+      type t = int
+      let dummy_value = let x = 3 in 0
+    end)
+
+[%%expect{|
+module type TInt = sig type t = int end
+val f : (module T : TInt) -> T.t -> T.t = <fun>
+Line 8, characters 28-29:
+8 |       let dummy_value = let x = 3 in 0
+                                ^
+Warning 26 [unused-var]: unused variable x.
+
+Line 8, characters 28-29:
+8 |       let dummy_value = let x = 3 in 0
+                                ^
+Warning 26 [unused-var]: unused variable x.
+
+val raise_duplicated_warning : int -> int = <fun>
```

Comment:
Typing happens twice :
- one for doing the module type substitution
- a second is the regular type checking for applications that also handles mode at that moment
(on the upstream version this test is raising the warning only once because this problem does not happen)